x-pack/plugins/alerts/server/authorization/alerts_authorization.ts

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import { pluck, mapValues } from 'lodash';
+import { KibanaRequest } from 'src/core/server';
+import { ALERTS_FEATURE_ID } from '../../common';
+import { AlertTypeRegistry } from '../types';
+import { SecurityPluginSetup } from '../../../security/server';
+import { RegistryAlertType } from '../alert_type_registry';
+import { PluginStartContract as FeaturesPluginStart } from '../../../features/server';
+import { AlertsAuthorizationAuditLogger, ScopeType } from './audit_logger';
+
+export interface RegistryAlertTypeWithAuth extends RegistryAlertType {
+  authorizedConsumers: string[];
+}
+
+export interface ConstructorOptions {
+  alertTypeRegistry: AlertTypeRegistry;
+  request: KibanaRequest;
+  features: FeaturesPluginStart;
+  auditLogger: AlertsAuthorizationAuditLogger;
+  authorization?: SecurityPluginSetup['authz'];
+}
+
+export class AlertsAuthorization {
+  private readonly alertTypeRegistry: AlertTypeRegistry;
+  private readonly features: FeaturesPluginStart;
+  private readonly request: KibanaRequest;
+  private readonly authorization?: SecurityPluginSetup['authz'];
+  private readonly auditLogger: AlertsAuthorizationAuditLogger;
+
+  constructor({
+    alertTypeRegistry,
+    request,
+    authorization,
+    features,
+    auditLogger,
+  }: ConstructorOptions) {
+    this.request = request;
+    this.authorization = authorization;
+    this.features = features;
+    this.alertTypeRegistry = alertTypeRegistry;
+    this.auditLogger = auditLogger;
+  }
+
+  public async ensureAuthorized(alertTypeId: string, consumer: string, operation: string) {
+    const { authorization } = this;
+    if (authorization) {
+      const alertType = this.alertTypeRegistry.get(alertTypeId);
```

Comment:
> how should an administrator deal with unknown alert types? 
That's a good question... they now won't be able to delete them, but tbh I don't know how we can get around this as we can't construct the _privilege action_ without the producer field, which we need to get from the registry 🤔 

We *could* do a _soft_ get from the registry so that we skip auth for producer level if the type is unknown - which would mean, as a super user, you'd get past that check and you'd be able to delete the alert.

Sine non-super users won't have privileges to that type either, only super users will be able to get past this check... so that's a possible solution... but it feels backwards.

> Is the alerting framework smart enough to exclude these now-unknown alert types from its queries? 
They won't come up in `find` as we filter out types your'e note authorized to read, but if there's an old task it will keep running until it reaches it's max failure run which I think is 5 time by default.

> Is it ok that these alerts can't be cleaned up?
Not ideal. but not a deal breaker. We might want to address this in a broader way... @mikecote should we add some issue around that?